### ISSUE TITLE
Cannot read property 'WebSocket' of undefined

### ISSUE THREAD

**Issue by soyjuanmedina** [socketio/engine.io#584]
*Note*: for support questions, please use one of these channels: [stackoverflow](http://stackoverflow.com/questions/tagged/engine.io) or [slack](https://socketio.slack.com)

### You want to:

* [x] report a *bug*
* [ ] request a *feature*

### Current behaviour
After update my Angular App to Angular 8 I can npm start and npm build, but when I upoload to the server the console of the browser yield 
`Cannot read property 'WebSocket' of undefined`

### Other information (e.g. stacktraces, related issues, suggestions how to fix)

I really don't know if the issue is for this library, but I read the problem is about the
`global.WebSocket || global.MozWebSocket;`
https://github.com/webpack/webpack-dev-server/issues/174#issuecomment-102391902
in some libraries and I use yours

**Comment by soyjuanmedina** [comment]
I really don't know who is guilty, but I supousse it's not your (but I can't delete the issue).

I solve it changing compilerOptions in tsconfig.json from 'es2015' to 'es5'

Thanks and sorry if not your fault.

**Comment by MickL** [comment]
I have the same problem.

I wonder why this error occurs because it is an important new feature of Angular 8 to compile modern js AND older js seperate if defined in browserlist. But only if you set to es2015 so it should not be changed back to es5.

**Comment by MickL** [comment]
@soyjuanmedina Could you please reopen?

**Comment by NicolasKritter** [comment]
@MickL 

> I think the more correct way to get `window` is `Function('return this')()` instead. The content in `Function` constructor is non-strict regardless of the environment it is created in.

_Originally posted by @andy0130tw in https://github.com/webpack/webpack-dev-server/issues/174#issuecomment-467771912_

 I had this problem after an upgrade to angular 8.1 and it works perfectly after doing what he suggested. 
I made a small patch.js to run after npm install in my package.json

```
const { readFileSync, writeFileSync,existsSync } = require('fs');
const patchIdentifier = '//-patched';

const fileToPatch = './node_modules/websocket/lib/browser.js';
const codeToPatch= 'var _global = (function() { return this; })();';
const patchCode = "var _global = Function('return this')();" + patchIdentifier;

function doPatch(fileName,sourceCode) {
  if (!existsSync(fileName)) {
    console.log('file not found '+fileName)
    process.exit(0);
  }
  const contents = readFileSync(fileName).toString().split('\n');
  // Check if code has been patched already
  const hasBeenPatched = contents.find(line => line.indexOf(patchIdentifier) !== -1);
  
  if (!hasBeenPatched) {
    const lineNumber = contents.findIndex(line => line.indexOf(sourceCode) !== -1);
    if (lineNumber < 0) {
      console.error('Could not find source code. Please check ' + fileName + ' and update the patch accordingly');
      process.exit(1);
    }
    // replace the line
    contents.splice(lineNumber, 1, patchCode);
    const updatedContents = contents.join('\n');
    writeFileSync(fileName, updatedContents);

    console.log('websocket browser.js has been Monkey patched');
  } else {
    console.log('websocket browser.js has already been patched');
  }
}
doPatch(fileToPatch,codeToPatch);
process.exit(0);
```
PS: with this solution, I could keep the es2015 in my tsconfig

**Comment by loveneetabsolvetech** [comment]
![Annotation 2019-12-31 133701](https://user-images.githubusercontent.com/57665672/71652520-3220db80-2d4c-11ea-83cf-612aab9ec31d.png)
Just need to update your opentok.js file as per screenshot

**Comment by MickL** [comment]
@loveneetabsolvetech Could you create a pull request? Or maybe someone of the contributors could fix this issue?

**Comment by darrachequesne** [comment]
I think this should be fixed by https://github.com/socketio/engine.io-client/commit/357f01d90448d8565b650377bc7cabb351d991bd (included in [3.4.1](https://github.com/socketio/engine.io-client/releases/tag/3.4.1)).

**Comment by Chingshangkhomba** [comment]
> ![Annotation 2019-12-31 133701](https://user-images.githubusercontent.com/57665672/71652520-3220db80-2d4c-11ea-83cf-612aab9ec31d.png) Just need to update your opentok.js file as per screenshot

this works
